### ISSUE TITLE
Set up @material/web for use in the plugin

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#5802]
The [@material/web](https://www.npmjs.com/package/@material/web) package should be added to Site Kit (behind a feature flag), as a starting point to switch to Material 3.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The [@material/web](https://www.npmjs.com/package/@material/web) package should be added to the Site Kit NPM dependencies.
* Following up to #5958, similar files should be scaffolded for a GM3 set of components:
    * A new `assets/js/googlesitekit/components-gm3/index.js` file should be created that exports a `Components` object, with the components as described in the next bullet point.
    * For now, files should be scaffolded for all the eventual GM3 components that already exist as GM2 versions in the corresponding `assets/js/googlesitekit/components-gm2` directory. The components in those files can simply be empty (return `null`) for now.
    * A new `assets/js/googlesitekit-components-gm3.js` entry point file should be created, as an equivalent to the similar `assets/js/googlesitekit-components-gm2.js` version, just for GM3 (i.e. using the above new `index.js` file.
* A new `material3` feature flag should be added.
* _If_ the `material3` feature flag is enabled, the JS asset `googlesitekit-components` (registered in PHP) should point to the `assets/js/googlesitekit-components-gm3.js` file instead of the `assets/js/googlesitekit-components-gm2.js` file.
* Storybook should be configured so that it is possible for different stories to rely on different versions of `googlesitekit-components`. A single story never needs to rely on both, but it should be possible for a story to specify that it should use the GM3 version of `googlesitekit-components`, without that breaking other stories that rely on the (default) GM2 version.

### Add @material/web NPM dependency

- `npm install @material/web`

### Create new GM3 module files

- Create `assets/js/googlesitekit/components-gm3/index.js` - this can be a direct copy of `assets/js/googlesitekit/components-gm2/index.js`.
- For each component file listed in `assets/js/googlesitekit/components-gm3/index.js`, create a stub file for it which exports a component by that name, rendering `null`.
- Create `assets/js/googlesitekit-components-gm3.js` - this can be a direct copy of `assets/js/googlesitekit-components-gm2.js`, except the import of `Components` should come from `./googlesitekit/components-gm3`.

### Add webpack config for googlesitekit-components-gm3 asset

Within `webpack.config.js` at the repo root:

- In the `webpackConfig` function, under where the `googlesitekit-components-gm2` entry point is defined, add a new GM3 entry point similar to the GM2 one: `'googlesitekit-components-gm3': './assets/js/googlesitekit-components-gm3.js'`.
- To prevent `@material/web` from being included in the `vendor` bundle, and instead make sure it will be included in the new `googlesitekit-components-gm3` bundle, locate the `optimization.splitChunks.cacheGroups.vendor` configuration and update the `test` parameter. The can be changed from a regex matching `[\\/]node_modules[\\/]` to a function which tests a) that the file is in `node_modules` and b) that it's _not_ in `@material/web`. In other words:

```js
test: ( module ) => {
	return (
		/[\\/]node_modules[\\/]/.test( module.resource ) &&
		! /[\\/]@material[\\/]web[\\/]/.test( module.resource )
	);
},
```

- The `@material/web` package is distributed as a suite of files to individually import and these files make use of the nullish coalescing operator `??` . As a result it's necessary to transpile these files. Within the `createRules` function, add another rule to the returned set of rules. This rule can be a copy of the existing rule with the following changes:
  - Update the `test` regex to match `node_modules/@material/web/.*.js`.
  - Remove the `exclude` parameter.

### Add the `gm3Components` feature flag

- Add `'gm3Components'` to `feature-flags.json`.

### Allow the `googlesitekit-components` PHP `Script` entry to switch between GM2 and GM3

- Within `includes/Core/Assets/Assets.php`, locate the `googlesitekit-components` `Script` entry. Change the static string `js/googlesitekit-components-gm2.js` to a conditional string based on whether the `gm3Components` feature flag is enabled, with the resulting value being either `js/googlesitekit-components-gm2.js` or `js/googlesitekit-components-gm3.js` accordingly.
- In order for the above to work, some additional changes are needed to the generated Manifest file and the `Script` class.
- At the moment each Manifest entry comprises a one-to-one mapping of handle to filename & hash pair. In the case of the `googlesitekit-components` entry this currently looks like this (obviously the hash value is liable to change):
  - `'googlesitekit-components' => array( "googlesitekit-components-gm2.js", "e9b9158e8c5a43f827c9" ),`
- Within `Scripts::register` there is logic that ensures the filename & hash are looked up from the Manifest file. However we now need to be able to conditionally choose from two different filenames.
- Taking the above two points in mind we need to do the following:
  - Within `webpack.config.js`, update the `generate` function returned by `manifestArgs` so that the resulting `googlesitekit-components` Manifest entry is an array of filename & hash pairs, one for GM2 and one for GM3, so it looks like this:
    - `'googlesitekit-components' => array( array( "googlesitekit-components-gm2.js", "e9b9158e8c5a43f827c9" ), array( "googlesitekit-components-gm3.js", "87814bbc2d79a9d8e5b9" ) )`
  - Update the `Script::register` method so that if it sees an array of arrays in the Manifest, it chooses which filename & hash pair to use based on which is a match for the filename part of the passed `$src` parameter.

### Update Storybook to allow switching between GM2/GM3 on a per-story basis

_Note: The Hot Module Reloading aspect of the Storybook setup was initially looked into for a solution here, but it ultimately looked too complex to get working within a reasonable amount of time (if at all). Instead, we take the following approach with a conditional export dependant on the presence of a query param:_

- Create a new entry point file for the components module in `.storybook/assets/js/googlesitekit-components.js`.
  - Within this file, check for the presence of the query param `isMaterial3`. Require either `../../../assets/js/googlesitekit-components-gm2` or `../../../assets/js/googlesitekit-components-gm3` depending on the presence of the query param and assign the result to `module.exports`.
- Within `.storybook/preview.js` add a new decorator to the end of the `decorators` list. This decorator should check the passed story `parameters` for the presence of an `isMaterial3` boolean property.
  - If `parameters.isMaterial3` is truthy, check for the presence of the `isMaterial3` query param. If the query param does not exist, set it and reload the page (updating `global.location`, so the iframe reloads when viewing Storybook in iframe mode, or the whole page reloads when showing the story in standalone mode).
  - If `parameters.isMaterial3` is falsy, check for the presence of the `isMaterial3` query param. If the query param does exist, unset it and reload the page.
- Update `.storybook/webpack.config.js`:
  - Where `siteKitPackageAliases` is created, in the callback to `mapValues` for the case where `api === 'components'`, replace `assets/js/googlesitekit-components-gm2.js` with the new `.storybook/assets/js/googlesitekit-components.js`.
  - Add the same rule to `config.module.rules` as was added to the main `webpack.config` to ensure `@material/web` files are transpiled.
- With the above in place, setting `parameters.isMaterial3` for a story will ensure the GM3 components are used, otherwise the GM2 components will continue to be used. The page will be reloaded as necessary which could have a bit of a knock-on effect on any VRT tests for GM3-enabled stories, where a delay may need to be introduced to account for the reload. Although this is slightly lacking in elegance it seems like a good enough solution considering we will eventually be removing the GM2 components, at which point we can also remove this code from Storybook.

### Test Coverage

* No new tests needed.

## QA Brief

- This doesn't really need any regular QA as a simple smoke test will be implicitly performed via the QA:Eng testing.

#### Create a test @material/web component

_Note, don't forget to `npm ci` :)_

As this PR sets up the infrastructure for components to use `@material/web`, but does not actually provide such a component, in order to verify it all works it's necessary to create a component that uses `@material/web`.

A good example is the `Button` component - edit `assets/js/googlesitekit/components-gm3/Button.js` and change the implementation to the following:

```js
import "@material/web/button/filled-button";

export default function Button() {
  return <md-filled-button label="Button"></md-filled-button>;
}
```

#### Check the GM3 button displays in Site Kit

- Run the build and check that `dist/manifest.php` has generated correctly with a nested array for the `googlesitekit-components` entry as detailed in the IB.
- Toggle the `gm3Components` feature flag.
- Instances of `Button` in the UI should switch to the GM3 implementation provided above.
- Test this for both production and development builds.

#### Check the GM3 button displays in Storybook

- In order to test in Storybook it's necessary to find a story that makes use of the `Button` imported from `googlesitekit-components`. An example could be the `ReminderBanner` stories.
- Add the `isMaterial3` flag to the parameters for one of the stories.
```js
Story.parameters = {
  isMaterial3: true,
};
```
- Check that the story reloads and displays the GM3 button for that story and switches back to GM2 for the others.
- Verify this for the story in the iframe and also the standalone story mode.

## Changelog entry

* N/A.

**Comment by techanvil** [comment]
Hi @felixarntz, having looked into the `@material/web` package, it strikes me that we may want to take a different approach to including it.

There is no main entry point for the package, we need to import the individual module for each web component that we want to use. So, taking the proposed approach, we'll need to create our own bundle for it, with an entry point JS module that includes all of the web components we want to use.

The drawback of this approach is we end up with a flat list of imports with no clear or inherent relationship between the web components and the React components which wrap them.

```js
// material-3.js

import '@material/web/badge/badge';
import '@material/web/button/filled-button';
import '@material/web/checkbox/checkbox';
import '@material/web/switch/switch';
```

I've done a bit of experimentation, creating React wrapper components which import their `@material/web` dependency within the React component file. What I'm seeing is the `@material/web` modules end up in the vendor bundle, which seems pretty OK to me.

```js
// Button.js

import '@material/web/button/filled-button';

export default function Button( { label } ) {
        return <md-filled-button label={ label }></md-filled-button>;
}
```

Also, while not exactly tree shaking, we can still avoid importing unused `@material/web` dependencies this way (in the PR linked below, note how `M3Badge` is defined but not used, and `@material/web/badge/badge` does not end up in the bundle).

The obvious downside is we'd lose the ability to conditionally include `@material/web` via a feature flag. But seeing as it will eventually be included unconditionally, and bearing in mind the module sizes don't seem too outlandish, I think the organisational benefit of importing the web components alongside their point of use would outweigh this conditional inclusion.

Here's a draft [PR](https://github.com/google/site-kit-wp/pull/5844) of the above, and the resulting bundle analysis. Please take a look and see what you think.

![image](https://user-images.githubusercontent.com/18395600/190214896-4d43f8ca-6b84-454c-a800-8f5d3ed4cb92.png)

![image](https://user-images.githubusercontent.com/18395600/190215135-a66d063e-8bf8-452c-898c-2985f53ecf51.png)

**Comment by felixarntz** [comment]
@techanvil Thank you for the detailed proposal.

The key thing I want to avoid is that we load the library in today's version of Site Kit. It should not be loaded and not be part of the ZIP production bundle as long as the users don't get any benefit from it - even more since it's not even stable yet.

As far as I understand your proposed approach, as long as the `@material/web` imports only happen in the new React component files which would not be used anywhere in the actual plugin for the foreseeable future, it should work - we just have to really avoid importing anything of that in the plugin (which should be fine as we're doing all the initial work around the components in Storybook, Jest etc.).

Still, there may be other tradeoffs to discuss, and if we prefer to keep the `@material/web` library in a separate "dist" file that should certainly be doable. I'm open to either approach as long as it satisfies the above.

Would love to hear additional feedback from @eugene-manuilov @aaemnnosttv @tofumatt

**Comment by techanvil** [comment]
Thanks @felixarntz.

> The key thing I want to avoid is that we load the library in today's version of Site Kit. It should not be loaded and not be part of the ZIP production bundle as long as the users don't get any benefit from it - even more since it's not even stable yet.

Thanks for articulating this point. Certainly makes sense to me.

> As far as I understand your proposed approach, as long as the `@material/web` imports only happen in the new React component files which would not be used anywhere in the actual plugin for the foreseeable future, it should work - we just have to really avoid importing anything of that in the plugin (which should be fine as we're doing all the initial work around the components in Storybook, Jest etc.).

That's right, if we took this approach we'd be able to avoid including `@material/web` in the bundle until we start using the new React components in the plugin.

Interested to hear any thoughts from @eugene-manuilov, @aaemnnosttv and/or @tofumatt on this one.

**Comment by eugene-manuilov** [comment]
@felixarntz @techanvil I think we should use the approach we discussed a while ago. The idea is to create a new entrypoint script for all common components that are separate from the plugin logic and can be used independently. We can call this script something like `googlesitekit-element2` and it will export all its components as `googlesitekit.element` global object.

We will move all common components (button, link, chip, etc) from the `components` folder into the new script and update all usages of moved components in the codebase to reference them via `googlesitekit-element` alias similarly how we do it with the `googlesitekit-data` script.

Once we have such script/entrypoint, we can work on the Material3 version of it where we will have the same components but implemented with the usage of the `@material/web` package. It will be a separate script/entrypoint in our webpack config, let's say `googlesitekit-element3`.

When both scripts ready, all we will need to do is to use the correct filename in the assets.php file to load the correct version of the element bundle. If the `material3` feature flag is enabled, then the `googlesitekit-element3` bundle will be used, otherwise `googlesitekit-element2` will be loaded.

**Comment by techanvil** [comment]
@eugene-manuilov this is an interesting idea, I can certainly see benefits to this approach. A few of points/questions:

- If we take this `googlesitekit-element` approach, we still need to choose how to include `@material/web`. And I think that importing the individual modules within their React wrappers would still be advantageous for the reasons I have detailed above. However `@material/web` should probably end up in the `googlesitekit-element` bundle(s) instead of the vendor bundle. This should be easy enough to achieve via webpack config.
- It raises a question about https://github.com/google/site-kit-wp/issues/5190 - should part of the AC be that it should be a drop-in replacement for the current `Checkbox` with the same props?

**Comment by eugene-manuilov** [comment]
> * If we take this `googlesitekit-element` approach, we still need to choose how to include `@material/web`. And I think that importing the individual modules within their React wrappers would still be advantageous for the reasons I have detailed above.

Yep, agree with a need to import individual modules within their React wrappers.

> * However `@material/web` should probably end up in the `googlesitekit-element` bundle(s) instead of the vendor bundle

I see no problems with it.

> * should part of the AC be that it should be a drop-in replacement for the current `Checkbox` with the same behaviour and props?

Yes, I think all components in `googlesitekit-element` should be as close as possible to the current versions of common components.

**Comment by felixarntz** [comment]
@eugene-manuilov I like your idea of the `googlesitekit-element` approach, that would be a clean approach. First of all, shouldn't it be `googlesitekit-components` though? `wp-element` is a wrapper for React, `wp-components` is similar to what we want to do here I think.

However, the effort to rewrite the entire codebase to import those components from `googlesitekit-element` rather than the `components` directory would probably be quite disruptive. Additionally, I would echo @techanvil's concern of keeping the component signatures matching - while we want to preferably do that, there may be cases where it's not possible, and then we would need to adjust the import code again.

I want to throw out another version of that idea for consideration: Could we keep all the current `components` untouched, and simply create a new `googlesitekit-components` asset for purely the Material 3 versions? This way we set up this cleaner system for the future, but we don't need to do all the extra work of retrofitting the existing codebase to that, just to get started. Of course that means there may be more work later in migrating the codebase to the components from the new asset, but it would also allow us to have minimal changes when needed. I'm also wary of making the extra work for Material 2 migration here a prerequisite because it delays when we can start building out the Material 3 components. What do you think @techanvil @eugene-manuilov?

**Comment by techanvil** [comment]
Thanks @felixarntz, that is also a good suggestion.

TL;DR: Sounds good to me, I'd be happy to take that approach.

---

Thinking it through a bit - say we take the approach that Eugene has suggested. It would be a non-trivial effort to move all of the current components into this new asset prior to starting the Material 3 work. However we could instead take an iterative approach - migrate one component at a time, move the component into the asset _and_ write the Material 3 version for it, before moving onto the next (obviously we can parallelize this with multiple devs).

This would have an advantage when the new component matches the old and it works as a drop-in replacement, because we wouldn't need a conditional import/render in each place where it's used, as the component version is managed at the asset dependency level.

However, in cases where the new component diverges from the old, we might need to update _all_ usages of it at once, potentially leading to some large PRs.

Taking the approach you have suggested would allow us to slice the work up a bit more finely, although it would obviously necessitate a bit more conditional importing or rendering.

I guess, on balance, I would lean toward the approach you have suggested on the basis that it would provide a bit more flexibility in how we manage the work.

As an aside, I must say, the more time I spend with `@material/web` components the more likely I think it is that we're going to find ourselves having to diverge a bit, here and there, from our current components. Maybe I am wrong - indeed I hope so, but the likelihood of running into an issue where we can't replicate behaviour due to some incompatibility in the exposed events for a web component, or simply due to some alpha-state glitch, feels like it's mounting up!

One last point - I am away for the next two weeks so someone else may wish to pick this up to progress it in the meantime. All the progress on the issue to date has been captured in this comment thread.

**Comment by felixarntz** [comment]
@eugene-manuilov @techanvil ACs here are updated based on the new approach (started in #5958), ready for IB!

**Comment by techanvil** [comment]
Hi @felixarntz, I think we should make a small change to the AC here.

Rather than initially scaffolding all the eventual GM3 components to return `null`, I think we should initially have `assets/js/googlesitekit/components-gm3/index.js` export the set of GM2 components.

That way we can integrate each component as we replace the GM2 version with the GM3 version, ensuring we can test any and all execution paths. Otherwise, we might find ourselves in a position where we can only fully test the integration of components with them all complete - a "big bang" approach where problems may arise that would have been caught earlier or avoided by a more iterative approach to the integration.

What do you think?

**Comment by felixarntz** [comment]
@techanvil Hmm, I'm not sure about how relevant that is given that we would not do any integrated testing of the new components until all of them have been implemented. They will only be used in Storybook, and we will have GitHub issues to implement the individual components. Until those are completed, we wouldn't even want to try loading the asset within Site Kit.

Maybe I'm misunderstanding something here, but from my perspective scaffolding the new components as empty is just a tiny bit closer to where we want to get to and shouldn't cause harm based on how we are planning out the overall epic. LMKWYT.

**Comment by techanvil** [comment]
@felixarntz, possibly I am being overly cautious, but I would rather not take this approach whereby we only load the asset in Site Kit once all the components are completed.

If we do run into issues at this point, by the nature of switching out lots of components all at once it would be a more complex scenario than if we were to integrate them individually. It strikes me that a more iterative approach to the integration would be less of a risk.

**Comment by felixarntz** [comment]
@techanvil We had discussed at a previous point that we would switch over the components at a single point in time. The next few weeks would only be spent in building out the components using Storybook, not integrate them in any way into the user-facing plugin side.

The current approach that we are taking, with there being two different versions of `googlesitekit-components`, would actually not even allow us do iteratively roll out those new components anyway. So in that case I don't even think it matters how we approach the components, as the plan was to build all of them out first, which means by definition we will switch to GM3 components only once they are all working.

We also have the feature flag, so in any case it wouldn't get rolled out to end users until we have spent sufficient time to be confident about the new components.

**Comment by techanvil** [comment]
@felixarntz, thanks, maybe I was a bit unclear in the point I was trying to make. I am not suggesting that we roll out the components iteratively - just that we integrate them iteratively behind the feature flag, so we catch any integration issues earlier that we might otherwise have.

Anyhow, as you've highlighted we do have the safety of the feature flag to ensure we can get things working properly before showing it to any users. So, I am not too concerned. Just thought it might be a slightly smoother path to development... Happy enough either way.

**Comment by felixarntz** [comment]
@techanvil Thanks for clarifying. In that case, I'm still slightly concerned about having a mix of the GM2+ and GM3 versions around together, that could also lead to problems in itself. For example, if a few of those components are related to each other and slightly differ in props / behavior.

IMO we can go with the approach to change them all at once eventually (behind the feature flag), as the feature flag should give us as much time as we need to catch regressions.

**Comment by techanvil** [comment]
Thanks @felixarntz - that sounds fine to me. Appreciate the point about having both versions around together, and how avoiding that would also be desirable.

As we were, then. :)

**Comment by tofumatt** [comment]
IB looks good to me. Anything specific can be worked out during code review.

Thanks for the context around the storybook reloading. 👍🏻

IB ✅

**Comment by hussain-t** [comment]
### QA:Eng Verified ✅ 

- Verified running the build and checked that `dist/manifest.php` had generated correctly with a nested array for the `googlesitekit-components` entry.
- Verified both production and development builds.
- Verified that the story reloads and displays the GM3 button, which has the `isMaterial3: true` parameter and switches back to GM2 for the others.
- Verified this for the story in the iframe and the standalone story mode.

<details>
<summary>Screenshots</summary>
<br>
<img width="1633" alt="Screenshot 2022-11-28 at 2 12 14 PM" src="https://user-images.githubusercontent.com/24408261/204266677-9bb9615e-bacd-4d47-b88b-f71027dc6620.png">

<img width="1620" alt="Screenshot 2022-11-28 at 2 12 27 PM" src="https://user-images.githubusercontent.com/24408261/204266685-ae37d5f8-825e-421d-b1bd-11684fd6b924.png">

<img width="805" alt="Screenshot 2022-11-28 at 2 12 39 PM" src="https://user-images.githubusercontent.com/24408261/204266707-0140e19a-a73c-41a9-9366-a96fbf22ba17.png">

<img width="1320" alt="Screenshot 2022-11-28 at 2 13 08 PM" src="https://user-images.githubusercontent.com/24408261/204266725-ce25680f-b7aa-473a-ba70-b78556560598.png">

<img width="1320" alt="Screenshot 2022-11-28 at 4 58 20 PM" src="https://user-images.githubusercontent.com/24408261/204266951-3c180765-64a6-4c9d-9201-b3210817db11.png">
</details>
